src/data/matrix/basic.lean

```diff
@@ -364,108 +377,96 @@ theorem diagonal_mul_diagonal' [decidable_eq n] (d₁ d₂ : n → α) :
 diagonal_mul_diagonal _ _
 
 @[simp]
-lemma map_mul {L : matrix m n α} {M : matrix n o α}
-  {β : Type w} [semiring β] {f : α →+* β} :
+lemma map_mul {L : matrix m n α} {M : matrix n o α} [semiring β] {f : α →+* β} :
   (L ⬝ M).map f = L.map f ⬝ M.map f :=
 by { ext, simp [mul_apply, ring_hom.map_sum], }
 
+lemma is_add_monoid_hom_mul_left (M : matrix l m α) :
+  is_add_monoid_hom (λ x : matrix m n α, M ⬝ x) :=
+{ to_is_add_hom := ⟨matrix.mul_add _⟩, map_zero := matrix.mul_zero _ }
+
+lemma is_add_monoid_hom_mul_right (M : matrix m n α) :
+  is_add_monoid_hom (λ x : matrix l m α, x ⬝ M) :=
+{ to_is_add_hom := ⟨λ _ _, matrix.add_mul _ _ _⟩, map_zero := matrix.zero_mul _ }
+
+protected lemma sum_mul (s : finset β) (f : β → matrix l m α)
+  (M : matrix m n α) : (∑ a in s, f a) ⬝ M = ∑ a in s, f a ⬝ M :=
+(@finset.sum_hom _ _ _ _ _ s f (λ x, x ⬝ M) M.is_add_monoid_hom_mul_right).symm
+
+protected lemma mul_sum (s : finset β) (f : β → matrix m n α)
+  (M : matrix l m α) : M ⬝ ∑ a in s, f a = ∑ a in s, M ⬝ f a :=
+(@finset.sum_hom _ _ _ _ _ s f (λ x, M ⬝ x) M.is_add_monoid_hom_mul_left).symm
+
+@[simp]
+lemma row_mul_col_apply (v w : m → α) (i j) : (row v ⬝ col w) i j = dot_product v w :=
+rfl
+
+end semiring
+
+section homs
+
 -- TODO: there should be a way to avoid restating these for each `foo_hom`.
 /-- A version of `one_map` where `f` is a ring hom. -/
-@[simp] lemma ring_hom_map_one [decidable_eq n]
-  {β : Type w} [semiring β] (f : α →+* β) :
+@[simp] lemma ring_hom_map_one [decidable_eq n] [semiring α] [semiring β] (f : α →+* β) :
   (1 : matrix n n α).map f = 1 :=
 one_map f.map_zero f.map_one
```

Comment:
These were previously in a section with  `[semiring α]`, which was wrong for most of the lemmas